encrypt/ipsec_rotate_key.go

```diff
@@ -0,0 +1,183 @@
+	}
+	return fmt.Sprintf("%d%s %s %s %s %s", k.spi, spiSuffix, k.algo, k.key, k.cipherMode, k.cipherKey)
+}
+
+var (
+	ipsecKeyRegex       = regexp.MustCompile(`^([[:digit:]]+\+?)[[:space:]](\S+)[[:space:]]([[:alnum:]]+)[[:space:]]([[:digit:]]+)$`)
+	cipherIPsecKeyRegex = regexp.MustCompile(`^([[:digit:]]+\+?)[[:space:]](\S+)[[:space:]]([[:alnum:]]+)[[:space:]](\S+)[[:space:]]([[:alnum:]]+)$`)
+	ipsecParsers        = map[*regexp.Regexp]func([]string) (ipsecKey, error){
+		ipsecKeyRegex:       keyFromSlice,
+		cipherIPsecKeyRegex: cipherKeyFromSlice,
+	}
+)
+
+func ipsecKeyFromString(s string) (ipsecKey, error) {
+	for matcher, parser := range ipsecParsers {
+		if matcher.MatchString(s) {
+			return parser(matcher.FindStringSubmatch(s))
+		}
+	}
+	return ipsecKey{}, fmt.Errorf("IPsec key has unsupported format")
+}
+
+func keyFromSlice(parts []string) (ipsecKey, error) {
+	if len(parts) != 5 {
+		return ipsecKey{}, fmt.Errorf("IPsec key invalid [expected parts: 5, actual parts: %d]", len(parts))
+	}
+	spiSuffix := false
+	if strings.HasSuffix(parts[1], "+") {
+		spiSuffix = true
+		parts[1] = strings.TrimSuffix(parts[1], "+")
+	}
+	spi, err := strconv.Atoi(parts[1])
+	if err != nil {
+		return ipsecKey{}, fmt.Errorf("invalid IPsec key SPI: %s", parts[1])
+	}
+	size, err := strconv.Atoi(parts[4])
+	if err != nil {
+		return ipsecKey{}, fmt.Errorf("invalid IPsec key size: %s", parts[4])
+	}
+	key := ipsecKey{
+		spi:       spi,
+		spiSuffix: spiSuffix,
+		algo:      parts[2],
+		key:       parts[3],
+		size:      size,
+	}
+	return key, nil
+}
+
+func cipherKeyFromSlice(parts []string) (ipsecKey, error) {
+	if len(parts) != 6 {
+		return ipsecKey{}, fmt.Errorf("IPsec key invalid [expected parts: 6, actual parts: %d]", len(parts))
+	}
+	spiSuffix := false
+	if strings.HasSuffix(parts[1], "+") {
+		spiSuffix = true
+		parts[1] = strings.TrimSuffix(parts[1], "+")
+	}
+	spi, err := strconv.Atoi(parts[1])
+	if err != nil {
+		return ipsecKey{}, fmt.Errorf("invalid cipher IPsec key SPI: %s", parts[1])
+	}
+	key := ipsecKey{
+		spi:        spi,
+		spiSuffix:  spiSuffix,
+		algo:       parts[2],
+		key:        parts[3],
+		cipherMode: parts[4],
+		cipherKey:  parts[5],
+	}
+	return key, nil
+}
+
+const maxIPsecSPI = 16
+
+func (k ipsecKey) rotate() (ipsecKey, error) {
+	key, err := generateRandom(len(k.key))
+	if err != nil {
+		return ipsecKey{}, fmt.Errorf("failed to generate authentication key: %s", err)
+	}
+
+	cipherKey := ""
+	if k.cipherMode != "" {
+		cipherKey, err = generateRandom(len(k.cipherKey))
+		if err != nil {
+			return ipsecKey{}, fmt.Errorf("failed to generate symmetric encryption key: %s", err)
+		}
+	}
+
+	spi := k.spi + 1
+	if spi >= maxIPsecSPI {
+		spi = 1
+	}
+	newKey := ipsecKey{
+		spi:        spi,
+		spiSuffix:  k.spiSuffix,
+		algo:       k.algo,
+		key:        key,
+		size:       k.size,
+		cipherMode: k.cipherMode,
+		cipherKey:  cipherKey,
+	}
+	return newKey, nil
+}
+
+func generateRandom(size int) (string, error) {
```

Comment:
Because this `size` is the length of a hex string, and not a number of bytes, this function is better named `generateRandomHex`. Otherwise, it's unclear that the `size` is a number of nybbles.

Reply:
Makes sense, renamed. Thank you!